### ISSUE TITLE
Exercise 20: You are an optimization freak

### ISSUE THREAD

**Issue by dag7dev** [sapienzastudentsnetwork/distributed-system-2023-2024#20]
Exercise 20: You are an optimization freak. You realize that, in some cases, it is not necessary that the proposer collects n-f’ (the Fast Paxos quorum) promises to take a decision. Which is the minimum quorum and under what hypothesis this minimum quorum is enough to take a decision?

**Comment by susannacifani** [comment]
I think the solution on Drive is correct:
Fast Paxos requires that any three fast roud quorums Qf intersect, i.e.
∀Q, Q0, Q00 ∈ Qf : Q ∩ Q0 ∩ Q00 6= ∅, meaning that a quorum of 2/3 of acceptors is
need to make safe choices. However, we can differentiate between phase-1 (prepare,
promise) quorums Q1 and phase-2 (accept, learn) fast round quorums Q2f . We
then observe that quorum intersection is only required between phase-1 quorums
Q1 and any pair of phase-2 fast round quorums Q2f , i.e. ∀Q ∈ Q1, ∀Q0, Q00 ∈ Q2f :
Q ∩ Q0 ∩ Q00 6= ∅. The weakened intersection requirements show that a simple
majority is sufficient for phase-1 of a fast round, instead of requiring 2/3 of the
acceptors.

**Comment by simonesestito** [comment]
But when is it correct? For every round or only the first one?

Also, could someone try to rephrase it? It's pretty strange to understand at a first glance 🤔

**Comment by lor3ny** [comment]
I've solved the exercise using the following reasoning (inspired by the drive solution), but i'm not sure if it is correct.

We can split fast paxos  execution in two macro phases:

1) Prepare-promise
2) Accept any-Accept

Phase 1 is handled only by the coordinator, so it works like a basic paxos because it doesn't allow other partecipant to request promises by the acceptors. That means the coordinator, in this phase, can follow the basic paxos quorum, so Max f: n = 2f +1.

Fast paxos, after the phase 1, send an "accept any" message. That message allows the acceptor to select and vote for the first accept message received by any proposer. This makes the proposers in Phase 2 concurrent, invalidanting the basic paxos quorum rule. Therefore, we need 2/3 quorum.

In summary: 1-phase(quorum>1/2) + 2-phase(quorum>2/3)

**Comment by simonesestito** [comment]
But are we allowed, in Fast Paxos, to send an _Accept-Any_ message **if and only if** it's the first round?

I mean: if, because of concurrency between multiple acceptors, no value is chosen and another round must be done, is the coordinator able to send another _Accept-Any_ message?
Is only 1/2 quorum enough in this situation?
